### ISSUE TITLE
Alias type fields naming inconsistencies

### ISSUE THREAD

**Issue by astefan** [elastic/elasticsearch#37141]
Considering an index with the following mapping:

```
        "parent": {
              "properties": {
                  "foo" : {
                    "type" : "keyword",
                    "ignore_above" : 1024
                },
                "foo2" : {
                    "type" : "text"
                }
              }
          },
          "nested": {
              "properties": {
                  "foo_alias2": {
                      "type":"alias",
                      "path":"parent.foo"
                }
              }
          },
          "nested.foo_alias1": {
              "type":"alias",
              "path":"parent.foo"
          },
          "nested.foo_alias2": {
              "type":"alias",
              "path":"parent.foo2"
          }
```

There are no errors when `PUT` `localhost:9200/test_alias_type` is run and the index is created successfully, but the following two issues are visible:

- `nested.foo_alias2` and `nested` -> `foo_alias2` can both be defined like this, even if one of them points to `parent.foo` while the other points to `parent.foo2`. Running a `query_string` on `nested.foo_alias2` (`"query_string": {"query": "nested.foo_alias2: xyz"}`) seems to indicate that `nested` -> `foo_alias2` definition won.
- `_field_caps` doesn't return a `nested` field of type `object`, only `nested.foo_alias2` is returned. This spawns two additional questions:
    - are `.` dots allowed in fields definitions for alias type of fields? If they are, shouldn't have the same behavior as a normal dotted field definition?

cc @jtibshirani

**Comment by astefan** [comment]
@jtibshirani thank you for the details.
There is still one unanswered question: 
> are . dots allowed in fields definitions for alias type of fields? If they are, shouldn't have the same behavior as a normal dotted field definition?

If I define this mapping:

```
        "properties" : {
          "parent": {
			  "type" : "keyword"
          },
          "a.b.c": {
          	"type":"alias",
          	"path":"parent"
          },
          "x.y.z": {
          	"type":"keyword"
          }
        }
```

and call `_field_caps` on the index, I get confusing a output:

```
        "parent": {
            "keyword": {
                "type": "keyword",
                "searchable": true,
                "aggregatable": true
            }
        },
        "x.y": {
            "object": {
                "type": "object",
                "searchable": false,
                "aggregatable": false
            }
        },
        "a.b.c": {
            "keyword": {
                "type": "keyword",
                "searchable": true,
                "aggregatable": true
            }
        },
        "x": {
            "object": {
                "type": "object",
                "searchable": false,
                "aggregatable": false
            }
        },
        "x.y.z": {
            "keyword": {
                "type": "keyword",
                "searchable": true,
                "aggregatable": true
            }
        }
```

`x.y.z` is "expanded" in a hierarchy of `object` fields, while the `alias` field `a.b.c` is not.

**Comment by jtibshirani** [comment]
Yes, dots are allowed in the names of field aliases, since they have the same structure as any other field (and can appear under an object mapping, etc.). I think what's happening in your example is the same as what I described above:

> @jimczi clarified offline that we follow the path of the concrete field when we resolve the hierarchy. So nested.foo_alias2 will check the parent.foo2 path and only include parent in the response.

When determining what object mappings to return, we first resolve alias fields to their concrete fields, then look at all parents of those concrete fields. This means that the object mappings `a` and `a.b` are never returned. It would be good to open a new issue to discuss this behavior, and if we think it should change.

**Comment by costin** [comment]
*tl;dr*

I've provided a workaround in SQL so there's no action that needs to be taken; I'm posting this mainly to record my findings in case somebody else bumps into this in the future. 
For the record, the approach taken in SQL is akin the mapping API - namely the structure for a dotted alias is visible
to the user.

*long story*

While working on the fix for #37224, I've with both the mapping and the field caps API and like @astefan I was surprised by the behavior between the two APIs in how they treat `object` mapping.

While the mapping API always reports the full hierarchy for a field regardless of its type (alias or not), `field_caps` does not.
Further more there's no property to indicate why this occurs (namely the target field is an alias) and one has to infer what's going by looking at the mapping API which provides information that does not appear in `field_caps`.

To wit, take the following mapping:
```json
{
  "mappings": {
    "doc": {
      "properties": {
        "x.y" : { "type" : "short" },
        "a.b" : { "type" : "alias" , "path" : "x.y" }
      }
    }
  }
}
```

Mapping API returns:
```
GET /an_alias/_mapping

{
  "an_alias": {
    "mappings": {
      "doc": {
        "properties": {
          "a": {
            "properties": {
              "b": {
                "type": "alias",
                "path": "x.y"
              }
            }
          },
          "x": {
            "properties": {
              "y": {
                "type": "short"
              }
            }
          }
        }
      }
    }
  }
}
```

which shows the object structure (notice `a` and `x`).

`field_caps` however provides a different view (I've eliminated fields starting with `_`):

```
GET /an_alias/_field_caps?fields=*

{
  "fields": {
    "x.y": {
      "short": {
        "type": "short",
        "searchable": true,
        "aggregatable": true
      }
    },
    "a.b": {
      "short": {
        "type": "short",
        "searchable": true,
        "aggregatable": true
      }
    },
    "x": {
      "object": {
        "type": "object",
        "searchable": false,
        "aggregatable": false
      }
    }
  }
}
```

notice there's `x` but there's no `a`.

I realize these are two different APIs: one is about mapping, the other one, fairly advanced and about fields. However the fact that fields of type `object` are *sometimes* returned in one vs *always* in the other is unexpected.

Asking for particular fields also has a peculiar behavior.
Take the response for a concrete field:

```
GET /an_alias/_field_caps?fields=x.y

{
  "fields": {
    "x.y": {
      "short": {
        "type": "short",
        "searchable": true,
        "aggregatable": true
      }
    },
    "x": {
      "object": {
        "type": "object",
        "searchable": false,
        "aggregatable": false
      }
    }
  }
}
```
`a.b` is not included in the pattern for `x.y` which is expected.

However for an alias field:

```
GET /an_alias/_field_caps?fields=a.b

{
  "fields": {
    "a.b": {
      "short": {
        "type": "short",
        "searchable": true,
        "aggregatable": true
      }
    },
    "x": {
      "object": {
        "type": "object",
        "searchable": false,
        "aggregatable": false
      }
    }
  }
}
```

`x` is being returned alongside `x.y`. I find this confusing since on one hand the initial alias name is preserved yet it's hierarchy is not.
(I think) I understand the reasons behind the behavior however as a user of both APIs I was quite surprised by the outcome. 
Some potential improvements can be:
1. return the hierarchy for alias fields, just like `mapping` does (so for `a.b` alias, there would be `a` returned as well)
2. signal that a field is an alias by specifying its target - this clarifies the rest of the structure which has a different naming then the potentially requested include:
```
GET /an_alias/_field_caps?fields=a.b

{
  "fields": {
    "a.b": {
      "short": {
        "type": "short",
        "searchable": true,
        "aggregatable": true,
        "alias" : "x.y"
      }
    },
    "x": {
      "object": {
        "type": "object",
        "searchable": false,
        "aggregatable": false
      }
    }
  }
}
```
3. explain the behavior for `alias` types in the docs (maybe the example in this ticket is good enough?)

Thanks,

**Comment by jimczi** [comment]
As Julie noticed in this [comment](https://github.com/elastic/elasticsearch/issues/37141#issuecomment-451527073), there are two issues outlined here. The first one about declaring a field with the same hierarchy in two ways should be tracked in a separate issue.
We discussed the second issue offline and agreed that we prefer to keep the alias field transparent in the `_field_caps` API. The `_field_caps` API is useful to merge multiple fields coming from different indices     so it makes sense to not treat alias fields differently (they can be alias in some indices and concrete fields in others). For these reasons I am closing this issue, we will keep the current behavior in `_field_caps` regarding aliases.

**Comment by jtibshirani** [comment]
As a note, I filed #37437 to discuss the question around duplicate field mappings.
